math/miller_rabin.cpp

```diff
@@ -0,0 +1,177 @@
+/**
+ * Copyright 2020 @author tjgurwara99
+ * @file
+ *
+ * A basic implementation of Miller-Rabin primality test.
+ */
+
+#include <cassert>
+#include <iostream>
+#include <random>
+#include <vector>
+
+/**
+ * Function to give a binary representation of a number in reverse order
+ * @param num integer number that we want to convert
+ * @return result vector of the number input in reverse binary
+ */
+template <typename T> std::vector<T> reverse_binary(T num) {
+    std::vector<T> result;
+    T temp = num;
+    while (temp > 0) {
+        result.push_back(temp % 2);
+        temp = temp / 2;
+    }
+    return result;
+}
+
+/**
+ * Function for modular exponentiation.
+ * This function is an efficient modular exponentation function.
+ * It can be used with any big integer library such as Boost multiprecision
+ * to give result any modular exponentation problem reletively quickly.
+ * @param base number being raised to a power as integer
+ * @param rev_binary_exponent reverse binary of the power the base is being
+ * raised to
+ * @param mod modulo
+ * @return r the modular exponentiation of $a^{n} \equiv r \mod{m}$ where $n$ is
+ * the base 10 representation of rev_binary_exponent and $m = mod$ parameter.
+ */
+template <typename T>
+T modular_exponentiation(T base, std::vector<T> rev_binary_exponent, T mod) {
```

Comment:
the second function argument should be passed in as:
`const std::vector<T> &rev_binary_exponenet`

Reply:
I haven't used std::vector<>::const_iterator so I was reluctant to use the value by const reference. Now I've used it but do look at the for loop if there is any issue that you see. It appears to work but just to be sure.